compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -294,17 +307,30 @@ trait BCodeBodyBuilder extends BCodeSkelBuilder {
         case t @ Try(_, _, _) =>
           generatedType = genLoadTry(t)
 
-        case Throw(expr) =>
-          generatedType = genThrow(expr)
+        case t: Apply if t.fun.symbol eq defn.throwMethod =>
+          generatedType = genThrow(t.args.head)
 
         case New(tpt) =>
-          abort(s"Unexpected New(${tpt.summaryString}/$tpt) reached GenBCode.\n" +
+          abort(s"Unexpected New(${tpt.tpe.showSummary()}/$tpt) reached GenBCode.\n" +
                 "  Call was genLoad" + ((tree, expectedType)))
 
-        case app @ Closure(env, call, functionalInterface) =>
+        case app: Closure =>
+          val env: List[Tree] = app.env
+          val call: Tree = app.meth
+          val functionalInterface: Symbol = {
+            val t = app.tpt.tpe.typeSymbol
+            if (t.exists) t
+            else {
+              val arity = app.meth.tpe.widenDealias.firstParamTypes.size - env.size
+              val returnsUnit = app.meth.tpe.widenDealias.resultType.classSymbol == defn.UnitClass
+              if (returnsUnit) ctx.requiredClass(("dotty.runtime.function.JProcedure" + arity))
+              else if (arity <= 2) ctx.requiredClass(("dotty.runtime.function.JFunction" + arity))
+              else ctx.requiredClass(("scala.Function" + arity))
+            }
+          }
           val (fun, args) = call match {
             case Apply(fun, args) => (fun, args)
-            case t @ Select(_, _) => (t, Nil)
+            case t @ DesugaredSelect(_, _) => (t, Nil)
```

Comment:
This can be a regular Select since the desugaring isn't used (and Ident is handled below)

Reply:
I would try to avoid any logic changes in this PR. To make sure that things work as before. We can optimize later.

Reply:
Added TODO comment